expression/collation.go

```diff
@@ -168,131 +192,283 @@ func deriveCoercibilityForColumn(c *Column) Coercibility {
-	return
-}
+func CheckAndDeriveCollationFromExprs(ctx sessionctx.Context, funcName string, evalType types.EvalType, args ...Expression) (et *ExprCollation, err error) {
+	ec := inferCollation(args...)
+	if ec == nil {
+		return nil, illegalMixCollationErr(funcName, args)
+	}
 
-// CheckAndDeriveCollationFromExprsWithCoer is the same with CheckAndDeriveCollationFromExprs, but also return Coercibility.
-func CheckAndDeriveCollationFromExprsWithCoer(ctx sessionctx.Context, funcName string, evalType types.EvalType, args ...Expression) (dstCharset, dstCollation string, coercibility Coercibility, err error) {
-	coll, chs, coercibility, legal := inferCollation(args...)
-	if !legal {
-		return "", "", CoercibilityIgnorable, illegalMixCollationErr(funcName, args)
+	if evalType != types.ETString && ec.Coer == CoercibilityNone {
+		return nil, illegalMixCollationErr(funcName, args)
 	}
 
-	if evalType != types.ETString && coercibility == CoercibilityNone {
-		return "", "", CoercibilityIgnorable, illegalMixCollationErr(funcName, args)
+	if evalType == types.ETString && ec.Coer == CoercibilityNumeric {
+		ec.Charset, ec.Collation = ctx.GetSessionVars().GetCharsetInfo()
+		ec.Coer = CoercibilityCoercible
+		ec.Repe = ASCII
 	}
 
-	if evalType == types.ETString && coercibility == CoercibilityNumeric {
-		charsetInfo, collation := ctx.GetSessionVars().GetCharsetInfo()
-		return charsetInfo, collation, CoercibilityCoercible, nil
+	if !safeConvert(ctx, ec, args...) {
+		return nil, illegalMixCollationErr(funcName, args)
+	}
+
+	return ec, nil
+}
+
+func safeConvert(ctx sessionctx.Context, ec *ExprCollation, args ...Expression) bool {
+	for _, arg := range args {
+		if arg.GetType().Charset == ec.Charset {
+			continue
+		}
+
+		if arg.Repertoire() == ASCII {
+			continue
+		}
+
+		if c, ok := arg.(*Constant); ok {
+			str, isNull, err := c.EvalString(ctx, chunk.Row{})
+			if err != nil {
+				return false
+			}
+			if !isNull && !isValidString(str, ec.Charset) {
+				return false
+			}
+		} else {
+			if arg.GetType().Collate != charset.CharsetBin && ec.Charset != charset.CharsetBin && !isUnicodeCollation(ec.Charset) {
+				return false
+			}
+		}
 	}
 
-	return chs, coll, coercibility, nil
+	return true
+}
+
+// isValidString check if str can convert to sdtChs charset without data lose.
+func isValidString(str string, dstChs string) bool {
+	switch dstChs {
+	case charset.CharsetASCII:
+		for _, c := range str {
+			if c >= 0x80 {
+				return false
+			}
+		}
+		return true
+	case charset.CharsetLatin1:
+		// For backward compatibility, we do not block SQL like select '啊' = convert('a' using latin1) collate latin1_bin;
+		return true
+	case charset.CharsetUTF8, charset.CharsetUTF8MB4:
+		// String in tidb is actually use utf8mb4 encoding.
+		return true
+	case charset.CharsetBinary:
+		// Convert to binary is always safe.
+		return true
+	default:
+		e, _ := charset.Lookup(dstChs)
+		_, err := e.NewEncoder().String(str)
```

Comment:
It is better to handle `e == nil` case.

Reply:
sorry, I did not get the point. what does the `err==nil` mean?